module-4/backend/week_three.md

```diff
@@ -1,19 +1,85 @@
-7. What are some advantages to having your front end and back end code live in separate applications? What are some disadvantages?
+
+  In app.js, we can import a route file and then tell the app which file to use for specific endpoints. It is not necessary to separate routes into different files, but it does improve organization.
+
+  Consider a resource `books` with a routes file /routes/books.js.
+    ```js
+      var books = require('./routes/books')
+      app.use('/books', books)
+    ```
+
+  Anytime the application receives a request to `/books`, it will look in the books route file. We'll add the route handling inside the `/routes/books.js` file.
+
+  ```js
+    var express = require('express')
+    var router = express.Router()
+
+    router.get('/', (req, res, next) => {
+      // code to get all Foods.
+    })
+  ```
+
+  Now the application can handle a GET request to the books endpoint, and we can add additional handling to return the specified resources.  
+
+4. What do we use Knex for?  
+
+  We use Knex to interact with the database to create and run migrations, write queries to retrieve records, and perform any other database operations.  
+
+  Knex has built in functions to query the database, but you can also write raw SQL.
+
+5. How **could** you organize your code to follow the MVC design pattern in your Quantified Self project?  
+
+  Following the MVC model in an express app would require creating directories for routes, controllers, and models.
+
+  The routes file would make a call to a controller action, and then the controller action would make calls to the model. With this organization, the model is the only layer that interacts directly with the database.  
+
+6. How do you execute raw SQL in node?  
+
+  Assuming that a database variable has been set at the top of the file, you can use Knex's `raw` function to write the SQL query.
+
+  ```js
+    return database.raw('SELECT * FROM foods;')
+    .then(foods => {
+      return foods.rows
+    })
+  ```
+
+7. What are some advantages to having your front end and back end code live in separate applications? What are some disadvantages?  
+
+  Some advantages to separating the front-end and back-end are increased modularity so that changes in a specific application does not affect the other application. The interaction is limited to the API endpoints. With a separate backend application, other applications could use the endpoints to fetch data.
+
+  A disadvantage to separating the applications is handling CORS requests, which means that a server has received a request from a different origin.  
 
 #### Review  
 
-8. Describe DNS.
-9. What does writing clean code mean to you?
-10. If you were building an application that models hotels, what classes would you need? What classes would be responsible for what functionality?
+8. Describe DNS.  
+
+  ?? Not confident in this answer at all, so this is without any additional research. ??  
```

Comment:
Thanks for this! We can also discuss in class to make sure everyone feels confident in their knowledge about this one.